### PR TITLE
Reproducibility to Improve Machine Learning

### PR DESCRIPTION
**Pull request by collins-a**
Article as index.md.
hero image as hero.jpg

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [ ] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by collins-a** [comment]
This fixes #894

File: articles/reproducibility/index.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+To aid in reproducibility, features that are implemented should not change after being created. They need to immutable. A new feature that is dependent on an already existing one should be represented separately. For example, it should have its own new column. Such measures help in the reduction of dependencies which in turn assists in achieving uniformity across environments. This in turn reduces the potential errors when duplicating a machine learning pipeline in contrasting environments.
+
+#### Model Building
+
+Algorithms themselves pose a challenge to reproducibility. As we mentioned before, many models have an element of randomness in their machine learning process. This randomness has models using the same training data giving different outcomes. For example, [cross validation](/engineering-education/evaluating-ml-model-performance/) may be reliant on random partitioning of data for folds.
+
+The potential solution to this is a simple one; keenness to order. Engineers should be keen on the order in which they pass features, hyperparameters as well as when to introduce [seed values](https://en.wikipedia.org/wiki/Initial_condition).
```
Comment (lalith1403) [inline articles/reproducibility/index.md]: Explain seeding as a way to ensure reproducibility. If required, insert a code snippet of numpy using this concept. np,random(seed=#).

File: articles/reproducibility/index.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+#### Algorithm
+
+Deterministic algorithms produce the same output given the same input, even on different iterations. However, there exist many algorithms that display non-determinism. Reproducibility in such algorithms is a difficult task since we expect randomness. Non-deterministic algorithms may be fed with the same input yet produce different results upon different iterations. The randomness in results may result from the algorithm using various paths to arrive at contrasting outcomes. A huge factor that influences the non-deterministic nature of an algorithm is the complexity of the model. The routes and processes used to achieve output may be non-reproducible, therefore ensuring an element of randomness. It is impossible to reproduce that which is random.
```
Comment (lalith1403) [inline articles/reproducibility/index.md:57]: How do we deal with non-deterministic algorithms? How do we ensure reproducibility in this case?

**Review by lalith1403 — CHANGES_REQUESTED** [review]
Amazing article. Elaborate on reproducibility under non-deterministic algorithms.

**Comment by collins-a** [comment]
Hey @lalith1403, thank you for the quality review. I have made the requested changes. Cheers!

**Comment by hectorkambow** [comment]
plagiarism check done

**Review by lalith1403 — APPROVED** [review]

**Pull request merged**